adapters/repos/db/lsmkv/segment_group_cleanup.go

```diff
@@ -529,13 +529,30 @@ func (sg *SegmentGroup) replaceSegment(segmentIdx int, tmpSegmentPath string,
 		return nil, fmt.Errorf("precompute segment meta: %w", err)
 	}
 
+	newSegment, err := sg.replaceSegmentBlocking(segmentIdx, oldSegment, precomputedFiles)
+	if err != nil {
+		return nil, fmt.Errorf("replace segment (blocking): %w", err)
+	}
+
+	if err := sg.deleteOldSegmentsNonBlocking(oldSegment); err != nil {
+		return nil, fmt.Errorf("delete old segments (non-blocking): %w", err)
+	}
```

Comment:
**question:** if it's non-blocking shall we just log instead of returning ?

Reply:
The non-blocking refers to "not holding an expensive lock", but it is still expected that the operation succeeds. I think the outcomes would be the following:

1. If we log and don't return the compaction is considered successful, but the old segment is never deleted. This does not block any operation but we do end up with an used file that would have to be manually deleted.
2. If we return the entire compaction is considered failed and it will be tried again the next iteration. If the delete fails again, we would be in an endless loop.

Based on that, I think I agree with you, I slightly prefer (1) as ending up with an unnecessary file on disk is the lesser of the two evils than blocking all future compactions for eternity if deleting the file consistently fails. Should hopefully be a very rare edge case.

Reply:
udpated.